vignettes/speedup_options.Rmd

```diff
@@ -0,0 +1,458 @@
+
+# Order by run time
+
+runtimes_dt <- runtimes_dt[order(runtime), ]
+
+# Print table
+knitr::kable(runtimes_dt, caption = "Run times for various _EpiNow2_ models")
+```
+
+### Estimation performance
+
+Now, we will compare the estimated and true values using the continuous ranked probability score (CRPS). The CRPS is a proper scoring rule that measures the accuracy of probabilistic forecasts. We will use the `crps()` function from the `{scoringutils}` package.
+
+To calculate the CRPS for the estimated R_t and infections, we will first set up a function that makes sure the true data and estimates are of the same length and calls the `crps_sample()` function from the `{scoringutils}` package.
+```{r crps-func}
+# A function to calculate the CRPS
+calc_crps <- function(x, truth) {
+  shortest_obs_length <- min(ncol(x), length(truth))
+  reduced_truth <- tail(truth, shortest_obs_length)
+  reduced_x <- tail(t(x), shortest_obs_length)
+  return(crps_sample(reduced_truth, reduced_x))
+}
+```
+
+Now, we will extract the $R_t$ and infection estimates and calculate the CRPS using the `calc_crps()` function above.
+
+```{r fit_crps}
+# Function to extract Rt estimates
+Rt_estimated <- lapply(results, function(x) {
+  if (is.null(x$error)) {
+    obj <- x$result$estimates$fit
+    if (inherits(obj, "stanfit")) {
+      if ("R[1]" %in% names(obj)) {
+        extract(obj, "R")$R
+      } else {
+        extract(obj, "gen_R")$gen_R
+      }
+    } else {
+      obj |>
+        as_draws_matrix() |>
+        subset_draws(variable = "R")
+    }
+  } else {
+    NA
+  }
+})
+# CRPS for the Rt estimates
+rt_crps <- lapply(
+  Rt_estimated,
+  calc_crps,
+  truth = R
+)
+
+# Function to extract infection estimates
+infections_estimated <- lapply(results, function(x) {
+  if (is.null(x$error)) {
+    obj <- x$result$estimates$fit
+    if (inherits(obj, "stanfit")) {
+      extract(obj, "infections")$infections
+    } else {
+      obj |>
+        as_draws_matrix() |>
+        subset_draws(variable = "infections")
+    }
+  } else {
+    NA
+  }
+})
+
+# CRPS for the infections estimates
+infections_crps <- lapply(
+  infections_estimated,
+  calc_crps,
+  truth = infections_true
+)
+```
+
+We will now post-process the CRPS results to make them easier to visualise by adding a  "time" column and reshaping the data to long format.
+
+```{r fit_postprocessing}
+# Post-processing the CRPS results of the Rt estimates
+rt_df <- as.data.table(rt_crps)
+rt_df[, time := 1:.N]
+rt_df <- melt(
+  rt_df,
+  id.vars = c("time"),
+  variable.name = "model"
+)
+
+# Post-processing the CRPS results of the infection estimates
+infections_df <- as.data.table(infections_crps)
+infections_df[, time := 1:.N]
+infections_df <- melt(
+  infections_df,
+  id.vars = c("time"),
+  variable.name = "model"
+)
+```
+
+Let's visualise the CRPS results for the $R_t$ and infection estimates.
+
+```{r rt_plot}
+rt_plot <- ggplot(rt_df, aes(x = time, y = value, colour = model)) +
```

Comment:
Maybe all of these should be relative to the default mode (as this is what people are using unless they do something else) and stratified by type (process model change or sampler change)